component/status/status.go

```diff
@@ -0,0 +1,194 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status // import "go.opentelemetry.io/collector/component/status"
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// EventType represents an enumeration of status event types
+type EventType int
+
+const (
+	// OK indicates the producer of a status event is functioning normally
+	OK EventType = iota
+	// RecoverableError is an error that can be retried, potentially with a successful outcome
+	RecoverableError
+	// PermanentError is an error that will be always returned if its source receives the same inputs
+	PermanentError
+	// FatalError is an error that cannot be recovered from and will cause early termination of the collector
+	FatalError
+)
+
+// Event is a status event produced by a component to communicate its status to registered listeners.
+// An event can signal that a component is working normally (i.e. Type: status.OK), or that it
+// is in an error state (i.e. Type: status.RecoverableError). An error status may optionally
+// include an error object to provide additional insight to registered listeners.
+type Event struct {
```

Comment:
I think we also need to know if this is a "receiver/processor/exporter"

Reply:
I added a `component.Kind` field to the event. Previously, the status reporting structs were in a `status` package (in `component/status`), but this forced me to move everything into component since this addition introduced an import cycle. Many things gained a `Status` prefix during this change.